### PR TITLE
[feat] Allow user to select any Google Font

### PR DESCRIPTION
**Pull request by a-wallen**
### Feature Preview

https://github.com/AppFlowy-IO/AppFlowy/assets/44445638/bcd8d676-954b-4e64-a36e-4bcd4e4108bc

fixes https://github.com/AppFlowy-IO/AppFlowy/issues/2795

---

#### PR Checklist

- [X] My code adheres to the [AppFlowy Style Guide](https://appflowy.gitbook.io/docs/essential-documentation/contribute-to-appflowy/software-contributions/submitting-code/style-guides)
- [X] I've listed at least one issue that this PR fixes in the description above.
- [x] I've added a test(s) to validate changes in this PR, or this PR only contains semantic changes.
- [x] All existing tests are passing.

### CODE REVIEW

**Comment by a-wallen** [comment]
@annieappflowy some of the fonts look _really_ bad 🤣

**Comment by a-wallen** [comment]
I just realized that the font style isn't plumbed into the editor. That needs to be changed before this lands.

**Comment by a-wallen** [comment]
Fixed the analyzer errors and realized that I couldn't implement `setMockInitialValues` for `FlowyPreferences` because the method is marked `@visibleForTesting`, so I had to allow `shared_preferences` as a dev_dependency. All calls should go to the wrapper class, but setting the mock values can be done directly through shared preferences. It will be a bit confusing, but I don't think there's any way around the issue.

**Review by MayurSMahajan — CHANGES_REQUESTED** [review]

File: frontend/appflowy_flutter/lib/workspace/application/appearance.dart [hunk 1]
```diff
@@ -49,6 +50,16 @@ class AppearanceSettingsCubit extends Cubit<AppearanceSettingsState> {
     emit(state.copyWith(themeMode: themeMode));
   }
 
+  /// Update selected font in the user's settings and emit an updated state
+  /// with the font name.
+  Future<void> setFontFamily(String fontFamilyName) async {
+    _setting.font = fontFamilyName;
+    _saveAppearanceSettings();
+    final prefs = await FlowyPreferences.instance;
+    await prefs.setString("kDocumentFontFamily", fontFamilyName);
```
Comment (a-wallen) [inline frontend/appflowy_flutter/lib/workspace/application/appearance.dart:59]: Use constant from `FlowyPreferences`

**Comment by annieappflowy** [comment]
It there a dynamic search as I type? This will let me quickly find a font

File: frontend/appflowy_flutter/packages/flowy_infra/lib/preferences/flowy_preferences.dart [hunk 2]
```diff
@@ -0,0 +1,86 @@
+import 'package:flutter/foundation.dart';
+import 'package:shared_preferences/shared_preferences.dart';
+
+typedef KeyValueNotifier = ValueNotifier<MapEntry<String, dynamic>>;
+
+class FlowyPreferences with ChangeNotifier implements SharedPreferences {
+  static const kDocumentAppearanceFontSize = 'kDocumentAppearanceFontSize';
```
Comment (LucasXu0) [inline frontend/appflowy_flutter/packages/flowy_infra/lib/preferences/flowy_preferences.dart:7]: I suggest moving these values to the `kv_keys.dart` file.

**Comment by a-wallen** [comment]
> It there a dynamic search as I type? This will let me quickly find a font

I'm back on this. Tried to fix the integration tests for a bit, but decided to go with a different approach. Will add the search ASAP.

**Review by hyj1204 — CHANGES_REQUESTED** [review]
https://github.com/AppFlowy-IO/AppFlowy/assets/14248245/be87c8b6-00c7-4ab2-aba8-cae021a72db6

I found these issues when I am running it.
1. The setting options(theme mode, font family and theme) didn't change. It changed after I restarted it.
2. The checkmark didn't get checked after clicking

I'm thinking about adding a "Default" option to the top of the font family selection. This option will be used to represent our default font, Poppins. With this option, users can easily switch back to the default font whenever they need to.
@annieappflowy @a-wallen

File: frontend/appflowy_flutter/lib/workspace/presentation/settings/widgets/settings_appearance_view.dart [hunk 3]
```diff
@@ -158,3 +134,147 @@ class ThemeModeSetting extends StatelessWidget {
...
+              return Column(
+                children: displayed
+                    .map(
+                      (font) => _fontFamilyItemButton(
+                        context,
+                        GoogleFonts.getFont(font),
+                      ),
+                    )
+                    .toList(),
+              );
```
[the remark this answers is not in the document]
Comment (a-wallen) [reply frontend/appflowy_flutter/lib/workspace/presentation/settings/widgets/settings_appearance_view.dart]: thanks 👍

**Comment by a-wallen** [comment]
> I found these issues when I am running it.
> 
> 1. The setting options(theme mode, font family and theme) didn't change. It changed after I restarted it.
> 2. The checkmark didn't get checked after clicking
> 
> I'm thinking about adding a "Default" option to the top of the font family selection. This option will be used to represent our default font, Poppins. With this option, users can easily switch back to the default font whenever they need to. @annieappflowy @a-wallen

All of these issues should be fixed. I think that having a button called `Reset` would be better; adding `Default` as an option in the dropdown is going to cause issues with sorted items in the list.
https://github.com/AppFlowy-IO/AppFlowy/assets/44445638/c9703f8a-f079-442e-bbbe-8015cc9dc561

File: frontend/appflowy_flutter/lib/workspace/presentation/settings/widgets/settings_appearance_view.dart [hunk 4]
```diff
@@ -158,3 +134,154 @@ class ThemeModeSetting extends StatelessWidget {
...
+                  context,
+                  GoogleFonts.getFont(displayed[index]),
+                ),
+                itemCount: displayed.length,
+                shrinkWrap: true,
+              );
+            },
+          ),
+        ],
+      ),
```
Comment (hyj1204) [inline frontend/appflowy_flutter/lib/workspace/presentation/settings/widgets/settings_appearance_view.dart:199]: I think using slivers here will improve the performance.
I also add a little padding on the right of textfield, so it won't be covered by the scrollbar.

image

```suggestion
      popupBuilder: (_) => CustomScrollView(
        shrinkWrap: true,
        slivers: [
          SliverPadding(
            padding: const EdgeInsets.only(right: 8),
            sliver: SliverToBoxAdapter(
              child: FlowyTextField(
                hintText: LocaleKeys.settings_appearance_fontFamily_search.tr(),
                autoFocus: false,
                debounceDuration: const Duration(milliseconds: 300),
                onChanged: (value) {
                  query.value = value;
                },
              ),
            ),
          ),
          const SliverToBoxAdapter(
            child: SizedBox(height: 4),
          ),
          ValueListenableBuilder(
            valueListenable: query,
            builder: (context, value, child) {
              var displayed = availableFonts;
              if (value.isNotEmpty) {
                displayed = availableFonts
                    .where(
                      (font) => font
                          .toLowerCase()
                          .contains(value.toLowerCase().toString()),
                    )
                    .sorted((a, b) => levenshtein(a, b))
                    .toList();
              }
              return SliverFixedExtentList.builder(
                itemBuilder: (context, index) => _fontFamilyItemButton(
                  context,
                  GoogleFonts.getFont(displayed[index]),
                ),
                itemCount: displayed.length,
                itemExtent: 32,
              );
            },
          ),
        ],
      ),
```

File: frontend/appflowy_flutter/lib/workspace/presentation/settings/widgets/settings_appearance_view.dart [hunk 5]
```diff
@@ -38,36 +50,18 @@ class ThemeSetting extends StatelessWidget {
...
-          },
+    return ThemeSettingDropDown(
+      label: LocaleKeys.settings_appearance_theme.tr(),
+      currentValue: currentTheme,
+      popupBuilder: (_) => IntrinsicHeight(
+        child: Column(
+          children: [
+            _themeItemButton(context, BuiltInTheme.defaultTheme),
+            _themeItemButton(context, BuiltInTheme.dandelion),
+            _themeItemButton(context, BuiltInTheme.lavender),
```
Comment (hyj1204) [inline frontend/appflowy_flutter/lib/workspace/presentation/settings/widgets/settings_appearance_view.dart:61]: same here. We can add MainAxisSize.min in Column and remove the IntrinsicHeight

**Review by hyj1204 — APPROVED** [review]
I have left some suggestions for changes, but overall, the rest of it looks good to me.

File: frontend/appflowy_flutter/lib/workspace/presentation/settings/widgets/settings_appearance_view.dart (hunk 5 above)
Comment (a-wallen) [reply frontend/appflowy_flutter/lib/workspace/presentation/settings/widgets/settings_appearance_view.dart:61]: Haha yeah this is much better :)

File: frontend/appflowy_flutter/lib/workspace/presentation/settings/widgets/settings_appearance_view.dart [hunk 6]
```diff
@@ -19,8 +24,15 @@ class SettingsAppearanceView extends StatelessWidget {
           return Column(
             crossAxisAlignment: CrossAxisAlignment.start,
             children: [
-              ThemeModeSetting(currentThemeMode: state.themeMode),
-              ThemeSetting(currentTheme: state.appTheme.themeName),
+              ThemeModeSetting(
```
Comment (MayurSMahajan) [inline frontend/appflowy_flutter/lib/workspace/presentation/settings/widgets/settings_appearance_view.dart:27]: Can we change the ordering here like this:
1. Theme
2. Theme Mode
3. Font Family
The logic behind it is that we want the theme and theme mode closer.
I don't know honestly, does it make sense?

**Review by MayurSMahajan — APPROVED** [review]

**Pull request merged**